daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -299,17 +299,26 @@ object Hash {
   // 1 - `templateId` is the identifier for a template with a key of type τ
   // 2 - `key` is a value of type τ
   @throws[HashingError]
-  def assertHashContractKey(templateId: Ref.Identifier, key: Value): Hash =
-    builder(Purpose.ContractKey, noCid2String)
-      .addIdentifier(templateId)
-      .addTypedValue(key)
-      .build
+  def assertHashContractKey(templateId: Ref.Identifier, key: Value, shared: Boolean): Hash = {
+    val hashBuilder = builder(Purpose.ContractKey, noCid2String)
+    (if (shared) {
+       hashBuilder.addQualifiedName(templateId.qualifiedName)
```

Comment:
In order to guarantee the scheme is collision free, it is safe to include an "empty" package ID. 
@andreaslochbihler-da do you agree ?

```suggestion
// The following is equivalent to adding an Identifier with empty package ID
// This guarantees there is no collision with no shared key,  as the package cannot be empty.
hashBuilder.add("").addQualifiedName(templateId.qualifiedName)
```

Reply:
You need to look at the bytes that are added to the hash builder.

For the record: 
* `hashBuilder.addIdentifier(templateId)` calls `hashBuilder.add(id.packageId).addQualifiedName(id.qualifiedName)`
* hashBuilder.add(string)` calls `hasBuilder.add(string.length).add(Utf8.bytes(string))`

So if you put the empty string, the bytes are `0x00000000` followed by `addQualifiedName(...)`.
If you put a non-empty package ID, the first four bytes will not be all zero because the string has a positive length. So the hashed bytes of a shared key and a non-shared key are never prefixes of each other and so we cannot run into hash collisions by misuse of the hash function.

In contrast, if you do not add the empty string, we need to dig a bit more:

* `hashBuilder.addQualfiiedName(name)` calls `addDottedName(name.module).addDottedName(name.name)`
* `addDottedName(name)` splits the dotted name into its components. Then it first adds four bytes for how many components there are and then the encoding of each component individually.

So to get a hash collision, you'd have as many components in the string as there are bytes in a package ID. I'd say that it is not completely impossible to have 32 dots in a dotted name. So without digging any further, it is conceivable that one might be able to construct a collision in the encoding here. Therefore, let's add the empty string.